tests/io/test_qualtran_io.py

```diff
@@ -0,0 +1,228 @@
+
+        w1, aux1 = bb.add(CNOT(), ctrl=w1, target=aux_wires[0])
+        w2, aux2 = bb.add(CNOT(), ctrl=w2, target=aux_wires[1])
+
+        ctrl_aux, w1 = bb.add(Toffoli(), ctrl=(aux1, aux2), target=w1)
+        ctrl_aux, w2 = bb.add(Toffoli(), ctrl=ctrl_aux, target=w2)
+        aux_wires = bb.join(ctrl_aux)
+
+        circuit_bloq = bb.finalize(wire1=w1, wire2=w2, aux_wires=aux_wires)
+
+        decomp = qml.FromBloq(circuit_bloq, wires=list(range(4))).decomposition()
+        expected_decomp = [
+            qml.H(0),
+            qml.H(1),
+            qml.CNOT([0, 2]),
+            qml.CNOT([1, 3]),
+            qml.Toffoli([2, 3, 0]),
+            qml.Toffoli([2, 3, 1]),
+        ]
+        assert decomp == expected_decomp
+
+        mapped_decomp = qml.FromBloq(circuit_bloq, wires=[3, 0, 1, 2]).decomposition()
+        mapped_expected_decomp = [
+            qml.H(3),
+            qml.H(0),
+            qml.CNOT([3, 1]),
+            qml.CNOT([0, 2]),
+            qml.Toffoli([1, 2, 3]),
+            qml.Toffoli([1, 2, 0]),
+        ]
+        assert mapped_decomp == mapped_expected_decomp
+        assert np.allclose(
+            qml.FromBloq(circuit_bloq, wires=list(range(4))).matrix(),
+            circuit_bloq.tensor_contract(),
+        )
+
+    def test_atomic_bloqs(self):
+        """Tests that atomic bloqs have the correct PennyLane equivalent after wrapped with `FromBloq`"""
+        from qualtran.bloqs.basic_gates import CNOT, Hadamard, Toffoli
+
+        assert Hadamard().as_pl_op(0) == qml.Hadamard(0)
+        assert CNOT().as_pl_op([0, 1]) == qml.CNOT([0, 1])
+        assert Toffoli().as_pl_op([0, 1, 2]) == qml.Toffoli([0, 1, 2])
+
+        assert qml.FromBloq(Hadamard(), 0).has_matrix is True
+        assert qml.FromBloq(Hadamard(), 0).decomposition() == []
+
+        assert np.allclose(qml.FromBloq(Hadamard(), 0).matrix(), qml.Hadamard(0).matrix())
+        assert np.allclose(qml.FromBloq(CNOT(), [0, 1]).matrix(), qml.CNOT([0, 1]).matrix())
+        assert np.allclose(
+            qml.FromBloq(Toffoli(), [0, 1, 2]).matrix(), qml.Toffoli([0, 1, 2]).matrix()
+        )
+
+    def test_bloqs(self):
+        """Tests that bloqs with decompositions have the correct PennyLane decompositions after
+        being wrapped with `FromBloq`"""
+
+        from qualtran.bloqs.basic_gates import Swap
+
+        assert qml.FromBloq(Swap(3), wires=range(6)).decomposition() == [
+            qml.SWAP(wires=[0, 3]),
+            qml.SWAP(wires=[1, 4]),
+            qml.SWAP(wires=[2, 5]),
+        ]
+
+        from qualtran.bloqs.basic_gates import ZPowGate
+        from qualtran.bloqs.phase_estimation import RectangularWindowState, TextbookQPE
+
+        textbook_qpe_small = TextbookQPE(ZPowGate(exponent=2 * 0.234), RectangularWindowState(3))
+
+        dev = qml.device("default.qubit")
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.FromBloq(textbook_qpe_small, wires=list(range(4)))
```

Comment:
Staring at this a bit longer made me realize `FromBloq` fails at one of the decomposed sub-Bloqs

```py
>>> qml.FromBloq(qt.bloqs.basic_gates.rotation.CZPowGate(0.468, eps=1e-11), wires=[3, 0]).decomposition()
WireError: Wires must be unique; got [3, 3, 0].
```

Reply:
Seems like for some reason, the bitsize for the register in `CZPowGate` returns `1` instead of `2` like it's supposed to.

See:
```
for r in CZPowGate(2, 1e-11).signature.rights():
    print(r.bitsize) # 1
```

Reply:
I can fix this bug but it shows an inconsistency in qualtran itself. When we look at the signature of CZPowGate itself, it suggests that only two qubits are needed. However, when we decompose CZPowGate, we'll see that actually, the decomposition requires a third ancillary qubit that's not in the signature.

Reply:
Hi @mpharrigan, I'm wondering if I've misunderstood something here. When we look at the signature for `CZPowGate`, it suggests that only two qubits are needed. However, in `build_composite_bloq()`, we clearly see that actually 3 qubits are needed (an additional ancillary qubit is needed for the And() operation). Is this expected behaviour? (Or should the signature include an additional qubit 'anc'?) 

For context, I'd like to know this so that I can know whether or not if CZPowGate should take in wires=[0, 1] or wires=[0, 1, 2].

Reply:
decompositions can include allocations (like `CZPowGate`). It allocates an ancilla and frees it by the end of the subroutine, so the signature shows two input/output wires

Reply:
Got it, thanks @mpharrigan! For PennyLane, operators/operations need to explicitly define all the wires that is required at the top level. In the case of something like `CZPowGate`, it would expect 3 wires. Is there some way in Qualtran to be able to look at a `Bloq` and figure this information out without necessarily going through the decomposition?